doc/tutorials/your_first_project/your_first_project.rst

```diff
@@ -0,0 +1,275 @@
+  }();
+  move_group_interface.setPoseTarget(target_pose);
+
+  // Create a plan to that target pose
+  auto const [success, plan] = [&move_group_interface]{
+    moveit::planning_interface::MoveGroupInterface::Plan msg;
+    auto const ok = static_cast<bool>(move_group_interface.plan(msg));
+    return std::make_pair(ok, msg);
+  }();
+
+  // Execute the plan
+  if(success) {
+    move_group_interface.execute(plan);
+  } else {
+    RCLCPP_ERROR(logger, "Planing failed!");
+  }
+
+3.1 Build and Run
+~~~~~~~~~~~~~~~~~
+
+Just like before we need to build the code before we can run it.
+
+In the workspace directory ``ws_moveit2`` run this command:
+
+.. code-block:: bash
+
+  colcon build --mixin debug
+
+After this succeeds, we need to re-use the demo launch file from the previous tutorial to start RViz and the MoveGroup node.
+In a separate terminal, source the workspace and then execute this:
+
+.. code-block:: bash
+
+  ros2 launch moveit2_tutorials demo.launch.py
+
+Then in the ``Displays`` window under ``MotionPlanning/Planning Request`` uncheck the box ``Query Goal State``.
+
+.. image:: rviz_1.png
+   :width: 300px
+
+In a third terminal source the workspace and run your program.
+
+.. code-block:: bash
+
+  ros2 run hello_moveit hello_moveit
+
+This should cause the robot in RViz to move and end up in this pose:
+
+.. image:: rviz_2.png
+   :width: 300px
+
+3.2 Examine the code
+~~~~~~~~~~~~~~~~~~~~
+
+The first thing we do is create the MoveGroupInterface.
+Note that this is the only mutable object (other than the thread) that we create in this program.
+
+.. code-block:: C++
+
+  using moveit::planning_interface::MoveGroupInterface;
+  auto move_group_interface = MoveGroupInterface(node, "panda_arm");
+
+Then we set our target pose and plan.
+
+.. code-block:: C++
+
+  // Set a target Pose
+  auto const target_pose = []{
+    geometry_msgs::msg::Pose msg;
+    msg.orientation.w = 1.0;
+    msg.position.x = 0.28;
+    msg.position.y = -0.2;
+    msg.position.z = 0.5;
+    return msg;
+  }();
+  move_group_interface.setPoseTarget(target_pose);
+
+  // Create a plan to that target pose
+  auto const [success, plan] = [&move_group_interface]{
+    moveit::planning_interface::MoveGroupInterface::Plan msg;
+    auto const ok = static_cast<bool>(move_group_interface.plan(msg));
+    return std::make_pair(ok, msg);
+  }();
+
+Finally we execute our plan if planning was successful, otherwise we log an error:
+
+.. code-block:: C++
+
+  // Execute the plan
+  if(success) {
+    move_group_interface.execute(plan);
+  } else {
+    RCLCPP_ERROR(logger, "Planing failed!");
+  }
+
+Summary
+-------
+
+You created a ROS 2 package and wrote your first program using MoveIt.
+
+
+Further Reading
+---------------
+
+- We used lambdas to be able to initialize objects as constants. This is known as a technique called IIFE.  `Read more about this pattern from C++ Stories <https://www.cppstories.com/2016/11/iife-for-complex-initialization/>`_.
```

Comment:
Maybe it's better to put this at the beginning of the tutorial... (or the first instance of this usage in the code via comment then refer to the footnote)? I was very confused by the use of all the lambdas until I got to here..

Also, while I don't really want to bikeshed this point, I do feel the readability of the code with the lambdas has gone down. The const-ness seems like a good thing, but I personally am not sure it is worth. Of course I'll defer this to you as I don't maintain this, just driving by :sweat_smile:

Reply:
That is a fair criticism.  I struggle with the tradeoffs of using advanced techniques like this or not here.  On one hand, I would like more of MoveIt itself (and our client projects) to use techniques like this to make the code better, on the other hand, I wonder if this technique will get in the way of a new user learning MoveIt.

The reason I settled on using it here is that this technique is used in modern C++ projects and is recommended by many of those that write about how to use C++.  If through these tutorials I can help new users write more modern C++ while they are learning MoveIt I want to do that.

The reason I want to keep this at the end is I think going and learning about this is a rabbit hole that I would rather not distract the user with.  Instead, I'd like them to see the syntax they find confusing but press on with the focus of learning MoveIt.  After they are done with the experience of this tutorial I've left these links so they can go learn about this pattern if they want to.